src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpDocumentExcerptService.cs

```diff
@@ -0,0 +1,192 @@
+            }
+
+            if (textSnapshot == null)
+            {
+                throw new ArgumentNullException(nameof(textSnapshot));
+            }
+
+            if (documentSnapshot is null)
+            {
+                throw new ArgumentNullException(nameof(documentSnapshot));
+            }
+
+            _mappingService = mappingService;
+
+            _textSnapshot = textSnapshot;
+            _documentSnapshot = documentSnapshot;
+        }
+
+        internal override async Task<ExcerptResultInternal?> TryGetExcerptInternalAsync(
+            Document document,
+            TextSpan span,
+            ExcerptModeInternal mode,
+            CancellationToken cancellationToken)
+        {
+            if (_textSnapshot == null)
+            {
+                return null;
+            }
+
+            var mapped = await _mappingService.MapSpansAsync(document, new[] { span }, cancellationToken).ConfigureAwait(false);
+            if (mapped.Length == 0 || mapped[0].Equals(default(RazorMappedSpanResult)))
+            {
+                return null;
+            }
+
+            var razorDocumentText = _documentSnapshot.Snapshot.AsText();
+            var razorDocumentSpan = razorDocumentText.Lines.GetTextSpan(mapped[0].LinePositionSpan);
+
+            var generatedDocument = document;
+
+            // First compute the range of text we want to we to display relative to the primary document.
+            var excerptSpan = ChooseExcerptSpan(razorDocumentText, razorDocumentSpan, mode);
+
+            // Then we'll classify the spans based on the primary document, since that's the coordinate
+            // space that our output mappings use.
+            var classifiedSpans = await ClassifyPreviewAsync(
+                razorDocumentSpan,
+                excerptSpan,
+                generatedDocument,
+                cancellationToken).ConfigureAwait(false);
+
+            // Now translate everything to be relative to the excerpt
+            var offset = 0 - excerptSpan.Start;
+            var excerptText = razorDocumentText.GetSubText(excerptSpan);
+            excerptSpan = new TextSpan(0, excerptSpan.Length);
+            razorDocumentSpan = new TextSpan(razorDocumentSpan.Start + offset, razorDocumentSpan.Length);
+
+            for (var i = 0; i < classifiedSpans.Count; i++)
+            {
+                var classifiedSpan = classifiedSpans[i];
+                var updated = new TextSpan(classifiedSpan.TextSpan.Start + offset, classifiedSpan.TextSpan.Length);
+                Debug.Assert(excerptSpan.Contains(updated));
+
+                classifiedSpans[i] = new ClassifiedSpan(classifiedSpan.ClassificationType, updated);
+            }
+
+            return new ExcerptResultInternal(excerptText, razorDocumentSpan, classifiedSpans.ToImmutable(), document, span);
+        }
+
+        private async Task<ImmutableArray<ClassifiedSpan>.Builder> ClassifyPreviewAsync(
```

Comment:
`ClassifyPreviewAsync` has a decent set of modifications. I'm working on refactoring out shared components but this will still need to be a separate function.

The way the previous excerpt service worked was it took the generated output, and then extracted the SourceMappings. Then it iterrated through the source mappings and did a bunch of changes around it.

This implementation uses the direct mapping(s) provided by the Span mapping service to find the classified preview.